caas/kubernetes/tunnel.go

```diff
@@ -0,0 +1,198 @@
+// Copyright 2020 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package kubernetes
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"math/rand"
+	"net"
+	"net/http"
+
+	"github.com/juju/errors"
+	corev1 "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	meta "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/kubernetes/scheme"
+	"k8s.io/client-go/rest"
+	"k8s.io/client-go/tools/portforward"
+	"k8s.io/client-go/transport/spdy"
+)
+
+// Tunnel represents an ssh like tunnel to a Kubernetes Pod or Service
+type Tunnel struct {
+	client     rest.Interface
+	config     *rest.Config
+	Kind       TunnelKind
+	LocalPort  string
+	Namespace  string
+	Out        io.Writer
+	readyChan  chan struct{}
+	RemotePort string
+	stopChan   chan struct{}
+	Target     string
+}
+
+type TunnelKind string
+
+const (
+	TunnelKindPods     = TunnelKind("pods")
+	TunnelKindServices = TunnelKind("services")
+)
+
+// Close disconnects a tunnel connection
+func (t *Tunnel) Close() {
+	close(t.stopChan)
+}
+
+func (t *Tunnel) findSuitablePodForService() (*corev1.Pod, error) {
+	clientSet := kubernetes.New(t.client)
+	service, err := clientSet.CoreV1().Services(t.Namespace).
+		Get(context.TODO(), t.Target, meta.GetOptions{})
+	if k8serrors.IsNotFound(err) {
+		return nil, errors.NewNotFound(err, "can't find service "+t.Target)
```

Comment:
errors.NotFoundf(err, "service %s", t.Target)

Reply:
This function signature doesn't exist